### PR TITLE
feat(gateway): add real-time notification system with SSE delivery

### PR DESCRIPTION
**Pull request by djdanielsson**
## Summary
- Add a complete notification system: backend event generation from scan results, persistent storage, REST CRUD API, real-time SSE delivery, and full frontend integration with the existing PatternFly notification drawer and toast alerts.
- Gitleaks secrets-detected notifications use danger variant with no auto-dismiss toast, per user requirement.

### Backend (Gateway)
- **DB model**: `Notification` ORM table with type, title, message, variant, project/scan FK (with `ondelete=CASCADE`), read state
- **SQLite FK enforcement**: `PRAGMA foreign_keys=ON` via engine connect event so cascade deletes work
- **CRUD queries**: `insert_notification`, `list_notifications` (deterministic ordering with id tie-breaker), `mark_notification_read`, `mark_all_notifications_read`, `delete_notification`
- **Notification generator** (`notifications.py`): Creates notifications from scan events — `scan_complete` (always), `secrets_detected` (SEC:* violations, danger variant), `health_changed` (score drops >= 10 points). Handles empty file paths gracefully. Catches only `DetachedInstanceError` for lazy-load safety.
- **SSE hub**: In-process fan-out via `asyncio.Queue` to connected browser clients, with 15s keep-alive heartbeat (non-cancelling `asyncio.wait` pattern) and proxy-friendly headers (`Cache-Control: no-cache`, `Connection: keep-alive`, `X-Accel-Buffering: no`)
- **Broadcast-after-commit**: `broadcast_notifications()` is a separate function called only after the DB transaction commits, preventing SSE clients from seeing notifications that fail to persist
- **REST endpoints**:
  - `GET /api/v1/notifications` — paginated, optional `unread_only` filter
  - `GET /api/v1/notifications/stream` — SSE `text/event-stream`
  - `PATCH /api/v1/notifications/{id}/read` — mark single as read
  - `POST /api/v1/notifications/read-all` — mark all as read
  - `DELETE /api/v1/notifications/{id}` — delete single
- **ReportingServicer integration**: `generate_notifications()` called after `ReportFixCompleted` DB commit

### Frontend
- `NotificationItem` TypeScript interface
- API client functions for all notification endpoints
- `useNotificationStream` hook: initial REST load + EventSource SSE subscription, pushes to `usePageNotifications` Zustand store, displays toast alerts via `usePageAlertToaster`
- Secrets-detected toasts: danger variant, no auto-dismiss
- Hook wired into `App.tsx` within `PageFramework` context

### Tests
- 22 unit tests covering notification generation, REST endpoints, SSE streaming delivery, broadcast-after-commit, edge cases (empty file paths), Gitleaks secrets detection, and health score drop scenarios

## Test plan
- [x] `tox -e lint` passes
- [x] `tox -e unit` passes (22 new tests pass; pre-existing failures in rule_doc_integration and session_venv_e2e are unrelated)
- [x] No new SDLC artifacts needed (notification system is an internal Gateway feature, no new services or protocol changes)

### CODE REVIEW

File: src/apme_gateway/notifications.py [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+            scan_id=scan.scan_id,
+            link=f"/projects/{scan.project_id}" if scan.project_id else f"/activity/{scan.scan_id}",
+        )
+        payloads.append(_notif_to_payload(notif_health))
+
+    # -- Broadcast all to connected SSE clients -----------------------------
+
+    for p in payloads:
+        _broadcast(p)
+
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_gateway/notifications.py]: Extracted broadcasting into a separate `broadcast_notifications()` function. Both callers (ReportingServicer and WebSocket handler) now commit the DB transaction first, then broadcast. This ensures SSE clients never see notifications that failed to persist. Fixed in 9518548.

File: src/apme_gateway/notifications.py [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+        List of notification payloads that were created and broadcast.
+    """
+    payloads: list[dict[str, Any]] = []
+
+    display_name = scan.project_path
+    try:
+        if scan.project is not None:
+            display_name = scan.project.name
+    except Exception:
+        pass
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_gateway/notifications.py:165]: Narrowed the catch to `DetachedInstanceError` (the specific SQLAlchemy exception for lazy-load on detached instances). Other real bugs will now propagate instead of being silently swallowed. Fixed in 9518548.

File: src/apme_gateway/notifications.py [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+        sec_files = sorted({v.file for v in sec_violations if v.file})
+        file_list = ", ".join(sec_files[:5])
+        if len(sec_files) > 5:
+            file_list += f" (+{len(sec_files) - 5} more)"
+
+        notif_sec = await insert_notification(
+            db,
+            type="secrets_detected",
+            title="Secrets Detected",
+            message=f"{display_name}: {len(sec_violations)} secret(s) found in {file_list}",
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_gateway/notifications.py]: Added a guard: when `sec_files` is empty (all violations have blank file paths), the message now renders as "N secret(s) found" without the trailing "in ". Added a dedicated test for this edge case. Fixed in 9518548.

File: src/apme_gateway/db/models.py [hunk 4]
```diff
@@ -478,3 +478,40 @@ class ScanGraph(Base):
...
+
+    __tablename__ = "notifications"
+
+    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
+    type: Mapped[str] = mapped_column(Text, nullable=False)
+    title: Mapped[str] = mapped_column(Text, nullable=False)
+    message: Mapped[str] = mapped_column(Text, nullable=False, default="")
+    variant: Mapped[str] = mapped_column(Text, nullable=False, default="info")
+    project_id: Mapped[str | None] = mapped_column(Text, ForeignKey("projects.id"), nullable=True)
+    scan_id: Mapped[str | None] = mapped_column(Text, ForeignKey("scans.scan_id"), nullable=True)
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_gateway/db/models.py]: Added `ondelete="CASCADE"` to both the `project_id` and `scan_id` FK columns on the Notification model. Deleting a project or scan will now cascade-delete its notifications at the DB level. Fixed in 9518548.

File: src/apme_gateway/api/router.py [hunk 5]
```diff
@@ -1968,6 +1969,118 @@ async def delete_rule_config(rule_id: str) -> None:
...
+    queue = subscribe()
+
+    async def _stream() -> AsyncIterator[str]:
+        try:
+            async for chunk in sse_event_stream(queue):
+                yield chunk
+        finally:
+            unsubscribe(queue)
+
+    return StreamingResponse(_stream(), media_type="text/event-stream")
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_gateway/api/router.py]: Added `test_sse_stream_delivers_notification`: opens the SSE stream via `client.stream()`, triggers a broadcast, and asserts the client receives the expected `data:` JSON payload with correct content-type header. Fixed in 9518548.

[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_gateway/api/router.py]: Added a 15-second keep-alive heartbeat (`: keep-alive` SSE comment) using `asyncio.wait_for` on `anext(stream)`, plus proxy-friendly response headers: `Cache-Control: no-cache`, `Connection: keep-alive`, `X-Accel-Buffering: no`. Fixed in 9518548.

File: src/apme_gateway/api/router.py [hunk 6]
```diff
@@ -1974,6 +1975,137 @@ async def delete_rule_config(rule_id: str) -> None:
...
+            while True:
+                try:
+                    chunk = await asyncio.wait_for(anext(stream), timeout=15.0)
+                except TimeoutError:
+                    yield ": keep-alive\n\n"
+                    continue
+                except StopAsyncIteration:
+                    break
+                yield chunk
+        finally:
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_gateway/api/router.py:2187]: Good catch — `asyncio.wait_for` cancels the awaitable on timeout, which kills the underlying async generator. Replaced with `asyncio.wait()` on a pending task: the task stays alive across timeouts, so the SSE connection persists through idle periods without dropping. Fixed in 3fafb69.

File: src/apme_gateway/db/queries.py [hunk 7]
```diff
@@ -1941,3 +1942,142 @@ async def project_graph(db: AsyncSession, project_id: str) -> ScanGraph | None:
...
+        Tuple of (notification rows, total count).
+    """
+    base = select(Notification)
+    if unread_only:
+        base = base.where(Notification.read.is_(False))
+
+    count_stmt = select(func.count()).select_from(base.subquery())
+    total = cast(int, (await db.execute(count_stmt)).scalar_one())
+
+    stmt = base.order_by(Notification.created_at.desc()).limit(limit).offset(offset)
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_gateway/db/queries.py]: Added `Notification.id.desc()` as a secondary sort key so ordering is deterministic even when timestamps collide. Fixed in 3fafb69.

File: src/apme_gateway/db/models.py [hunk 8]
```diff
@@ -484,3 +484,44 @@ class ScanGraph(Base):
...
+        read: Whether the user has marked this notification as read.
+    """
+
+    __tablename__ = "notifications"
+
+    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
+    type: Mapped[str] = mapped_column(Text, nullable=False)
+    title: Mapped[str] = mapped_column(Text, nullable=False)
+    message: Mapped[str] = mapped_column(Text, nullable=False, default="")
+    variant: Mapped[str] = mapped_column(Text, nullable=False, default="info")
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_gateway/db/models.py:524]: Added a `_set_sqlite_pragma` listener on the engine's `connect` event that executes `PRAGMA foreign_keys=ON` for every new connection, so `ondelete=CASCADE` is enforced at the SQLite level. Fixed in 3fafb69.

File: frontend/src/hooks/useNotificationStream.ts [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+}
+
+const NO_DISMISS_TYPES = new Set(['secrets_detected']);
+
+export function useNotificationStream(): void {
+  const { setNotificationGroups } = usePageNotifications();
+  const alertToaster = usePageAlertToaster();
+  const mountedRef = useRef(true);
+  const itemCacheRef = useRef<NotificationItem[]>([]);
+
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply frontend/src/hooks/useNotificationStream.ts:77]: Removed `itemCacheRef` entirely — it was written to on initial load and every SSE message but never read. Fixed in 0b4b8c7.

File: tests/test_gateway_notifications.py [hunk 10]
```diff
@@ -0,0 +1,493 @@
...
+            assert resp.status_code == 200
+            assert resp.headers["content-type"] == "text/event-stream; charset=utf-8"
+            async for line in resp.aiter_lines():
+                if line.startswith("data:"):
+                    data = json.loads(line[len("data: "):])
+                    assert data["type"] == "scan_complete"
+                    assert data["title"] == "Test"
+                    break
+
+
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply tests/test_gateway_notifications.py]: Wrapped the SSE read loop in `asyncio.wait_for(..., timeout=2.0)` so the test fails fast instead of hanging indefinitely. Also replaced `asyncio.get_event_loop().create_task()` with `asyncio.create_task()` to avoid the deprecation warning. Fixed in 0b4b8c7.

File: src/apme_gateway/api/router.py [hunk 11]
```diff
@@ -1974,6 +1975,146 @@ async def delete_rule_config(rule_id: str) -> None:
...
+
+    queue = subscribe()
+
+    async def _stream() -> AsyncIterator[str]:
+        stream = sse_event_stream(queue)
+        pending_chunk: asyncio.Task[str] | None = None
+        try:
+            while True:
+                if pending_chunk is None:
+                    pending_chunk = asyncio.ensure_future(anext(stream))  # type: ignore[arg-type]
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_gateway/api/router.py]: Replaced `asyncio.ensure_future(anext(stream))` with `asyncio.create_task(_next_chunk())` via a small wrapper coroutine. This removes the `type: ignore` and makes the intent clearer. Fixed in 0b4b8c7.

File: src/apme_gateway/notifications.py [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+        payload: JSON-serialisable notification dict.
+    """
+    dead: list[asyncio.Queue[dict[str, Any]]] = []
+    for q in _subscribers:
+        try:
+            q.put_nowait(payload)
+        except asyncio.QueueFull:
+            dead.append(q)
+    for q in dead:
+        _subscribers.remove(q)
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_gateway/notifications.py]: Replaced the drop-subscriber pattern with drop-oldest-item backpressure. On `QueueFull`, the oldest queued notification is discarded and the new one is enqueued, so slow consumers stay connected and receive newer events instead of becoming permanently stuck. Fixed in 081a311.

File: frontend/src/hooks/useNotificationStream.ts [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+      markNotificationRead(item.id).catch(() => {});
+    };
+
+    es.onerror = () => {
+      // EventSource auto-reconnects — nothing to do
+    };
+
+    return () => {
+      mountedRef.current = false;
+      es.close();
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply frontend/src/hooks/useNotificationStream.ts]: Restructured `useNotificationStream` to `await` the initial REST load before opening the EventSource. The SSE connection now only starts after `buildGroups()` has populated the store, so incoming SSE messages merge into existing state instead of being overwritten. Fixed in 081a311.

File: src/apme_gateway/notifications.py [hunk 14]
```diff
@@ -0,0 +1,249 @@
...
+        if scan.project is not None:
+            display_name = scan.project.name
+    except DetachedInstanceError:
+        pass
+
+    # -- Scan complete notification (always) --------------------------------
+
+    if scan.scan_type == "remediate":
+        title = "Remediation Complete"
+        msg = f"{display_name}: {scan.fixed_count} findings resolved, {scan.total_violations} remaining"
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_gateway/notifications.py]: Agreed — `total_violations` is the original total (`remaining + fixed`). Now computing remaining as `max(scan.total_violations - scan.fixed_count, 0)` and using that in the message. Test updated to seed `total_violations=7, fixed_count=5` and assert `"2 remaining"`. Fixed in d3fcc13.

File: tests/test_gateway_notifications.py [hunk 15]
```diff
@@ -0,0 +1,504 @@
...
+            scan = (await db.execute(select(Scan).where(Scan.scan_id == "scan-1"))).scalar_one()
+            payloads = await generate_notifications(db, scan, [])
+            await db.commit()
+
+        assert payloads[0]["type"] == "scan_complete"
+        assert payloads[0]["title"] == "Remediation Complete"
+        assert "5 findings resolved" in payloads[0]["message"]
+        assert payloads[0]["variant"] == "success"
+
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply tests/test_gateway_notifications.py:131]: Updated — test now seeds `total_violations=7, fixed_count=5` (total >= fixed) and asserts `"5 findings resolved"` + `"2 remaining"`, matching the actual engine semantics where `total = remaining + fixed`. Fixed in d3fcc13.

File: src/apme_gateway/api/router.py [hunk 16]
```diff
@@ -2231,7 +2376,25 @@ async def _run_op() -> tuple[str, object, str]:
                 )
                 if captured_patches:
                     await q.store_patches(db, completed_scan_id, captured_patches)
-                await q.update_project_health(db, proj.id)
+                old_hs = proj.health_score
+                new_hs = await q.update_project_health(db, proj.id)
+
+                scan_row = await q.get_scan(db, completed_scan_id)
+                if scan_row is not None:
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_gateway/api/router.py]: Fixed — health score is now computed directly from `scan_row.violations` via `compute_health_score()` and persisted on the project object in the same transaction, bypassing the `update_project_health` re-query. This eliminates the race where a concurrent scan could alter the health computation. The notification's `old_health_score`/`new_health_score` are now guaranteed to be tied to `completed_scan_id`. Fixed in bdb797c.

File: frontend/src/hooks/useNotificationStream.ts [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+        // Gateway may be unavailable — silent degradation
+        if (!mountedRef.current) return;
+      }
+
+      if (!mountedRef.current) return;
+
+      const proto = window.location.protocol === 'https:' ? 'https:' : 'http:';
+      const sseUrl = `${proto}//${window.location.host}/api/v1/notifications/stream`;
+      es = new EventSource(sseUrl);
+
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply frontend/src/hooks/useNotificationStream.ts:90]: Fixed — the EventSource now opens *before* the REST load. SSE events that arrive during the REST fetch are buffered, then merged with the REST response using id-based dedupe before setting state. Also added a duplicate guard in `mergeNotification` so even late-arriving SSE items that duplicate REST data are skipped. This closes the subscription gap entirely. Fixed in bdb797c.

File: src/apme_gateway/notifications.py [hunk 18]
```diff
@@ -0,0 +1,248 @@
...
+        try:
+            q.put_nowait(payload)
+        except asyncio.QueueFull:
+            with contextlib.suppress(asyncio.QueueEmpty):
+                q.get_nowait()
+            try:
+                q.put_nowait(payload)
+            except asyncio.QueueFull:
+                logger.warning("Dropping notification for persistently full subscriber queue")
+
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_gateway/notifications.py]: Fixed — `_broadcast` now iterates over `tuple(_subscribers)` (a snapshot), so concurrent `unsubscribe()` calls during iteration can't mutate the list mid-loop. Fixed in 60cb917.

File: frontend/src/hooks/useNotificationStream.ts [hunk 19]
```diff
@@ -0,0 +1,171 @@
...
+            markNotificationRead(buffered.id).catch(() => {});
+          }
+        }
+      } catch {
+        // Gateway may be unavailable — silent degradation
+        if (!mountedRef.current) return;
+      }
+
+      restLoaded = true;
+    };
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply frontend/src/hooks/useNotificationStream.ts:226]: Fixed — `restLoaded` is now set to `true` immediately after the REST response arrives (before draining the buffer), and `buffer.splice(0)` atomically captures all pending items. Any events arriving via `onmessage` after that point go through `handleSseItem` directly. This closes the window where late events could get stuck in the buffer. Fixed in 60cb917.

File: src/apme_gateway/api/router.py [hunk 20]
```diff
@@ -2231,7 +2376,28 @@ async def _run_op() -> tuple[str, object, str]:
                 )
                 if captured_patches:
                     await q.store_patches(db, completed_scan_id, captured_patches)
-                await q.update_project_health(db, proj.id)
+                scan_row = await q.get_scan(db, completed_scan_id)
+                old_hs = proj.health_score
+                if scan_row is not None:
+                    new_hs = q.compute_health_score(list(scan_row.violations))
+                    proj.health_score = new_hs
+
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_gateway/api/router.py:2474]: Fixed. Now fetches the Project row in the active session via `select(Project).where(Project.id == proj.id)` instead of using the detached `proj` instance. The health score update is properly persisted before commit. Fixed in bc35ff3.

File: src/apme_gateway/api/router.py [hunk 21]
```diff
@@ -1974,6 +1975,150 @@ async def delete_rule_config(rule_id: str) -> None:
...
+    return StreamingResponse(
+        _stream(),
+        media_type="text/event-stream",
+        headers={
+            "Cache-Control": "no-cache",
+            "Connection": "keep-alive",
+            "X-Accel-Buffering": "no",
+        },
+    )
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_gateway/api/router.py:2204]: Added `test_sse_endpoint_delivers_event` which opens `/api/v1/notifications/stream` via `client.stream()`, verifies SSE headers (content-type, cache-control, x-accel-buffering), broadcasts a payload, and asserts the client receives a `data:` event. Fixed in bc35ff3.

File: tests/test_gateway_notifications.py [hunk 22]
```diff
@@ -0,0 +1,486 @@
...
+            payload = {"id": 1, "type": "scan_complete", "title": "Test"}
+            _broadcast(payload)
+            chunk = await asyncio.wait_for(anext(stream), timeout=2.0)
+            data = json.loads(chunk.removeprefix("data: ").strip())
+            assert data["type"] == "scan_complete"
+            assert data["title"] == "Test"
+        finally:
+            unsubscribe(q)
+
+
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply tests/test_gateway_notifications.py:457]: Added a new `test_sse_endpoint_delivers_event` test that exercises the full endpoint via `client.stream()`, checks SSE headers and proxy headers, and validates broadcast delivery. The original unit-level test is kept for direct `sse_event_stream()` coverage. Fixed in bc35ff3.

File: frontend/src/hooks/useNotificationStream.ts [hunk 23]
```diff
@@ -0,0 +1,176 @@
...
+              actionClose: undefined,
+            });
+            markNotificationRead(buffered.id).catch(() => {});
+          }
+        }
+      } catch {
+        // Gateway may be unavailable — silent degradation
+        restLoaded = true;
+        if (!mountedRef.current) return;
+      }
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply frontend/src/hooks/useNotificationStream.ts]: Fixed. The catch path now drains the SSE buffer with id-based dedupe, pushes items into the notification drawer via `buildGroups()`, and fires toast alerts for each buffered item — so early SSE events are no longer silently dropped when REST is unavailable. Fixed in bc35ff3.

File: src/apme_gateway/grpc_reporting/servicer.py [hunk 24]
```diff
@@ -434,3 +436,43 @@ def _add_graph(db: AsyncSession, scan_id: str, content_graph_json: str) -> None:
...
+                level="",
+                message="",
+                file=v.file,
+            )
+            for v in all_protos
+        ]
+        payloads = await generate_notifications(db, scan, stub_violations)
+        await db.commit()
+        broadcast_notifications(payloads)
+    except Exception:
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_gateway/grpc_reporting/servicer.py:477]: Added `await db.rollback()` in the except block before the warning log so the session is left in a clean state for any subsequent operations. Fixed in bc35ff3.

File: tests/test_gateway_notifications.py [hunk 25]
```diff
@@ -0,0 +1,515 @@
...
+            assert resp.headers["content-type"].startswith("text/event-stream")
+            assert resp.headers.get("cache-control") == "no-cache"
+            assert resp.headers.get("x-accel-buffering") == "no"
+
+            payload = {"id": 1, "type": "scan_complete", "title": "SSE Test"}
+            _broadcast(payload)
+
+            lines = resp.aiter_lines()
+            for _ in range(20):
+                line = await asyncio.wait_for(anext(lines), timeout=2.0)
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply tests/test_gateway_notifications.py]: Good catch. Replaced the per-iteration 2s timeout with a single 5s overall deadline so the test fails fast instead of taking up to ~40s. Fixed in e5044fd.

File: frontend/src/hooks/useNotificationStream.ts [hunk 26]
```diff
@@ -0,0 +1,192 @@
...
+        const merged = [...resp.items];
+        const pending = buffer.splice(0);
+        for (const buffered of pending) {
+          if (!restIds.has(buffered.id)) {
+            merged.unshift(buffered);
+          }
+        }
+
+        setNotificationGroups(() => buildGroups(merged));
+
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply frontend/src/hooks/useNotificationStream.ts:177]: Fixed. Moved `restLoaded = true` after the `setNotificationGroups` call so the store is fully populated before SSE events start going through `handleSseItem`. Also switched from a full-replace to a prev-based merge updater so any notifications inserted concurrently by `handleSseItem` are preserved rather than overwritten. Fixed in 5494f34.

File: src/apme_gateway/api/router.py [hunk 27]
```diff
@@ -13,12 +13,14 @@
...
-from collections.abc import Sequence
+from collections.abc import AsyncIterator, Sequence
 from typing import cast
 
 from fastapi import APIRouter, HTTPException, Query, WebSocket
-from fastapi.responses import JSONResponse
+from fastapi.responses import JSONResponse, StreamingResponse
 from pydantic import BaseModel, Field
+from sqlalchemy import select
+from starlette.websockets import WebSocketDisconnect  # type: ignore[import-not-found]
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_gateway/api/router.py]: Added `import contextlib` to the top-level imports. Fixed in bb564dd.

File: frontend/src/hooks/useNotificationStream.ts [hunk 28]
```diff
@@ -0,0 +1,211 @@
...
+                if (!base[key].notifications.some((existing) => existing.id === n.id)) {
+                  base[key].notifications.unshift(n);
+                }
+              }
+            }
+          }
+          return base;
+        });
+        restLoaded = true;
+
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply frontend/src/hooks/useNotificationStream.ts:177]: Restructured the buffer drain: `restLoaded` is now flipped before processing buffered items, and a second drain pass catches any SSE events that arrived during the REST→store merge. Applied the same pattern in the catch (REST-failure) path. Fixed in bb564dd.

File: src/apme_gateway/api/router.py [hunk 29]
```diff
@@ -2055,7 +2059,445 @@ async def delete_rule_config(rule_id: str) -> None:
...
+                pending_chunk = None
+                yield chunk
+        finally:
+            if pending_chunk is not None:
+                pending_chunk.cancel()
+                with contextlib.suppress(asyncio.CancelledError, StopAsyncIteration):
+                    await pending_chunk
+            with contextlib.suppress(AttributeError):
+                await stream.aclose()  # type: ignore[attr-defined]
+            unsubscribe(queue)
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_gateway/api/router.py:2194]: Changed `sse_event_stream` return type from `AsyncIterator[str]` to `AsyncGenerator[str, None]` so `aclose()` is type-safe. Removed the `# type: ignore[attr-defined]` from both the router and the test. Fixed in 88829f5.

File: src/apme_gateway/api/router.py [hunk 30]
```diff
@@ -10,15 +10,18 @@
...
-from collections.abc import Sequence
+from collections.abc import AsyncIterator, Sequence
 from typing import cast
 
 from fastapi import APIRouter, HTTPException, Query, WebSocket
-from fastapi.responses import JSONResponse
+from fastapi.responses import JSONResponse, StreamingResponse
 from pydantic import BaseModel, Field
+from sqlalchemy import select
+from starlette.websockets import WebSocketDisconnect  # type: ignore[import-not-found]
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_gateway/api/router.py]: Switched to `from fastapi import WebSocketDisconnect` to match the rest of the gateway (e.g. `session_client.py`). This drops the blanket `# type: ignore[import-not-found]`. Fixed in 88829f5.

File: frontend/src/hooks/useNotificationStream.ts [hunk 31]
```diff
@@ -0,0 +1,235 @@
...
+          key: `notif-${item.id}`,
+          title: item.title,
+          children: item.message,
+          variant: item.variant,
+          timeout,
+          actionClose: undefined,
+        });
+
+        markNotificationRead(item.id).catch(() => {});
+      };
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply frontend/src/hooks/useNotificationStream.ts:105]: This is a fair observation, but the current design is intentional: the server-side `markNotificationRead` call persists read state so it survives page reloads and is consistent across tabs/devices. The vendored PatternFly notification drawer's local `readMap` is a UI-layer concern that tracks which notifications the user has *seen* during the current session (controls badge count / drawer styling). These are two different concepts — "server-acknowledged" vs "UI-session-seen" — and syncing them tightly would couple the hook to the framework's internal store API (`usePageNotificationsRead`) in a way that's fragile across framework upgrades.

That said, if users report that the badge stays lit after dismissing toasts, wiring `setNotificationRead` into the handler would be a reasonable follow-up. Leaving this unresolved for a human reviewer to weigh in on the UX trade-off.

File: src/apme_gateway/db/queries.py [hunk 32]
```diff
@@ -1956,3 +1957,142 @@ async def project_graph(db: AsyncSession, project_id: str) -> ScanGraph | None:
...
+        Tuple of (notification rows, total count).
+    """
+    base = select(Notification)
+    if unread_only:
+        base = base.where(Notification.read.is_(False))
+
+    count_stmt = select(func.count()).select_from(base.subquery())
+    total = cast(int, (await db.execute(count_stmt)).scalar_one())
+
+    stmt = base.order_by(Notification.created_at.desc(), Notification.id.desc()).limit(limit).offset(offset)
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_gateway/db/queries.py]: Refactored `list_notifications()` to build the count query directly on the `Notification` table with filters applied via `.where(*filters)` instead of wrapping the full `select(Notification)` in a subquery. Fixed in 9217d5e.

File: tests/test_gateway_notifications.py [hunk 33]
```diff
@@ -0,0 +1,498 @@
...
+        from apme_gateway.api.router import notification_stream
+
+        resp = await notification_stream()
+
+        assert resp.media_type == "text/event-stream"
+        assert resp.headers.get("Cache-Control") == "no-cache"
+        assert resp.headers.get("X-Accel-Buffering") == "no"
+        assert resp.headers.get("Connection") == "keep-alive"
+
+
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply tests/test_gateway_notifications.py:457]: Added `await resp.body_iterator.aclose()` after the header assertions so the generator's `finally` block runs and the subscriber is unregistered from the global `_subscribers` list. Fixed in 9217d5e.

File: src/apme_gateway/notifications.py [hunk 34]
```diff
@@ -0,0 +1,248 @@
...
+def subscribe() -> asyncio.Queue[dict[str, Any]]:
+    """Register a new SSE client and return its queue.
+
+    Returns:
+        An asyncio.Queue that will receive notification payloads.
+    """
+    q: asyncio.Queue[dict[str, Any]] = asyncio.Queue(maxsize=256)
+    _subscribers.append(q)
+    return q
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_gateway/notifications.py:68]: Acknowledged, but this is by design for the current deployment model. The Gateway runs as a single container in the Podman pod (ADR-012: scale pods as a unit, not individual services). There are no multi-worker or multi-instance Gateway deployments in the reference architecture. Adding Redis pub/sub or another cross-process mechanism would introduce a new infrastructure dependency for a scaling scenario that doesn't exist today.

If multi-worker Gateway deployment becomes a requirement, it should be captured as a new ADR and REQ — not bolted onto this PR. Leaving unresolved for a human reviewer.
